src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,87 +77,295 @@ def __eq__(self, other: Any) -> bool:
-            NotImplementedError: Raised when implementation not present
-
+            ValueError : Raised when invalid security_type is provided.
         """
         if len(session.parties) != 3:
             raise ValueError("Falcon requires 3 parties")
 
         result = None
 
         if session.protocol.security_type == "semi-honest":
-            args = list(zip(x.share_ptrs, y.share_ptrs))
+            result = Falcon.mul_semi_honest(x, y, session, op_str, **kwargs_)
+        elif session.protocol.security_type == "malicious":
+            result = Falcon.mul_malicious(x, y, session, op_str, **kwargs_)
+        else:
+            raise ValueError("Invalid security_type for Falcon multiplication")
 
-            z_shares_ptrs = parallel_execution(
-                Falcon.compute_zvalue_and_add_mask, session.parties
-            )(args)
+        result = ABY3.truncate(result, session)
 
-            result = MPCTensor(shares=z_shares_ptrs, session=x.session)
-            result.shape = MPCTensor._get_shape("mul", x.shape, y.shape)  # for prrs
-            result = ABY3.truncate(result, session)
+        return result
 
-        else:
-            raise NotImplementedError(
-                f"mult operation not implemented for {session.protocol.security_type} setting"
-            )
+    @staticmethod
+    def mul_semi_honest(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        truncate: bool = True,
+        **kwargs_: Dict[Any, Any],
+    ) -> MPCTensor:
+        """Falcon semihonest multiplication.
+
+        Performs Falcon's mul implementation, add masks and performs resharing.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            truncate (bool) : Applies truncation on the result if set.
+            kwargs_ (Dict[Any, Any]): Kwargs for some operations like conv2d
+
+        Returns:
+            MPCTensor: Result of the operation.
+        """
+        result = None
+
+        args = [
+            [x_share, y_share, op_str]
+            for x_share, y_share in zip(x.share_ptrs, y.share_ptrs)
+        ]
+
+        z_shares_ptrs = parallel_execution(
+            Falcon.compute_zvalue_and_add_mask, session.parties
+        )(args, kwargs_)
+
+        result = MPCTensor(shares=z_shares_ptrs, session=x.session)
+
+        if not truncate:
+            z_shares = [share.get() for share in z_shares_ptrs]
 
+            # Convert 3-3 shares to 2-3 shares by resharing
+            reshared_shares = ReplicatedSharedTensor.distribute_shares(
+                z_shares, x.session
+            )
+            result = MPCTensor(shares=reshared_shares, session=x.session)
+        result.shape = MPCTensor._get_shape(op_str, x.shape, y.shape)  # for prrs
         return result
 
+    @staticmethod
+    def triple_verification(
```

Comment:
Q: Shouldn't we do hear something like this?
![image](https://user-images.githubusercontent.com/7805588/124677325-1d308580-deb8-11eb-8d37-6db40973ccc2.png)
And in the `malicious_multiplication` reconstruct this value and if it is ```0 (zero)``` we know that the result (formed from the shares from "z") is correct.

Reply:
George for that part, semi -honest multiplication induces additional randomness by `PRZS` , this also is for beaver triple calculation, `PRZS` randomizes it.So we require the `reconstruction` step.

I considered several scenario's while working on it, there are scenario's where `PRZS` is required (just only semi-honest mult and without truncation) and `PRZS` is  not required when(semi-honest mult without truncation) , the same for malicious multiplication.

There are lot of optimizations which could be done,I thought to do it for another PR, after completion of basic functionalities, as the requirements might change then.

Reply:
I modified the reconstruction according to ABY3.